articles/20240819_developing_inside_containers.md

```diff
@@ -0,0 +1,290 @@
+
+   ```Docker
+
+file
+   RUN apt-get update
+   RUN apt-get install -y git
+   ```
+
+   **After:**
+
+   ```Dockerfile
+   RUN apt-get update && apt-get install -y git
+   ```
+
+### How Daytona Does It
+
+At Daytona, developing inside containers using VSCode’s Remote - Containers extension has been a game-changer for our workflow. Here’s how we do it:
+
+1. **Standardized Environments:** We use a `devcontainer.json` in every project to define the development environment. This ensures that all developers are working in the same environment, regardless of their local machine setup.
+
+   **Example:**
+
+   ```json
+   {
+     "name": "Daytona Dev Container",
+     "image": "daytona/base-image:latest",
+     "extensions": [
+       "dbaeumer.vscode-eslint",
+       "ms-python.python"
+     ],
+     "postCreateCommand": "npm ci"
+   }
+   ```
+
+2. **Custom Docker Images:** We create custom Docker images tailored to our project’s needs, including all necessary tools, libraries, and configurations. This helps us avoid "works on my machine" scenarios.
+
+   **Special Instruction:**
+
+   To create a custom Docker image, start with a base image, install your project's dependencies, and then push it to a container registry (like Docker Hub or a private registry).
+
+   **Example:**
+
+   ```Dockerfile
+   FROM node:14
+
+   WORKDIR /usr/src/app
+
+   COPY package*.json ./
+
+   RUN npm ci
+
+   COPY . .
+
+   CMD ["npm", "start"]
+   ```
+
+   Build and push the image:
+
+   ```bash
+   docker build -t daytona/base-image:latest .
+   docker push daytona/base-image:latest
+   ```
+
+3. **Automated Setup:** Our `devcontainer.json` files are designed to automate as much of the setup process as possible. For instance, we use the `postCreateCommand` to install dependencies automatically, ensuring that developers can start coding immediately after the container is built.
+
+   **Example:**
+
+   ```json
+   "postCreateCommand": "npm ci && npm run setup"
+   ```
+
+   This command not only installs dependencies but also runs a setup script that configures the environment according to project-specific needs.
+
+4. **Consistent Development Experience:** By using VSCode’s Remote - Containers extension, we ensure that our development experience is consistent across different platforms, making it easier for new developers to onboard and start contributing to projects.
+
+### Conclusion
+
+Developing inside containers with VSCode is a powerful way to ensure consistency and portability in your development workflow. By following the steps outlined in this guide, you can set up a containerized development environment that suits your needs, avoid common pitfalls, and take inspiration from how Daytona does it to streamline your process. Happy coding!
+
+---
+
+This article should provide a comprehensive, beginner-friendly guide to developing inside containers with VSCode, complete with examples and special instructions to help new developers avoid common issues.
```

Comment:
Mate what is this? 🤣

Reply:
I wrote that to be the header but I'll take it out🤣